### PR TITLE
i386: Add basic implementations of PGE and PAE, and also enabled relevant feature bits on later CPUs. [Melody Goad]

### PR DESCRIPTION
**Pull request by qeeg**

### CODE REVIEW

File: src/devices/cpu/i386/i386.cpp [hunk 1]
```diff
@@ -285,15 +381,30 @@ bool i386_device::translate_address(int pl, int type, uint32_t *address, uint32_
...
+				*error = ((type & TRANSLATE_WRITE) ? 2 : 0) | ((m_CPL == 3) ? 4 : 0);
+				if (entry)
+					*error |= 1;
+				return false;
+			}
+			vtlb_dynload(index, *address, entry);
+			return true;
 		}
-		vtlb_dynload(index, *address, entry);
-		return true;
```
Comment (cuavas) [inline src/devices/cpu/i386/i386.cpp:296]: Why did you duplicate the `vtlb_dynload(index, *address, entry); return true;` and move it inside both branches of the `if`/`else`?  In fact, most of the code inside the `if`/`else` is duplicated.  One of the foundational principles of good programming is DRY.

File: src/devices/cpu/i386/i386.cpp [hunk 2]
```diff
@@ -1728,8 +1845,12 @@ uint8_t i386_device::read8_debug(uint32_t ea, uint8_t *data)
...
-		return 0;
+	if (!(m_cr[4] & 0x20))
+	{
+		if (!i386_translate_address_normal(TRANSLATE_DEBUG_MASK,&address,nullptr))
+			return 0;
+	}
+	else if (!i386_translate_address_pae(TRANSLATE_DEBUG_MASK,&address,nullptr))
 
 	address &= m_a20_mask;
 	*data = m_program->read_byte(address);
```
Comment (cuavas) [inline src/devices/cpu/i386/i386.cpp:1856]: I don’t know if it’s intentional, but having the controlled statement of the `else if` at the same indentation level and separated from it by a blank line is like intentionally making the code difficult to understand.

**Review by cuavas — COMMENTED** [review]
The code is sloppy, and you opened the PR in a state that didn’t build.

How well has this actually been tested?  What software running in MAME exercises it?  I’d be very reluctant to merge untested code for a CPU core that’s as widely used as this.  The i386 core is used by projects outside MAME, including otvdm, so changes can impact a lot of people.

You don’t have a great track record here.  Your last set of changes to i386 (hardware breakpoints) didn’t work, and needed to be fixed by @Paul-Arnold.

**Comment by qeeg** [comment]
Many Linux distros require PAE nowadays, even if they're 32-bit compatible.

**Comment by cuavas** [comment]
> Many Linux distros require PAE nowadays, even if they're 32-bit compatible.

Answer the question.  How well have you tested the changes?

**Comment by qeeg** [comment]
I have not, but they're based on the documentation on sandpile.org and on working code from PCBox/86Box.

**Comment by cuavas** [comment]
It’s not worth merging an unexercised change to a widely-used CPU code.  Come back when you’re serious about proper software development.

File: src/devices/cpu/i386/i386.cpp [hunk 3]
```diff
@@ -3430,7 +3542,9 @@ void pentium4_device::device_reset()
 	// [ 0:0] FPU on chip
 	// [ 8:8] CMPXCHG8B instruction
 	// [15:15] CMOV and FCMOV
-	m_feature_flags = 0x00008101;       // TODO: enable relevant flags here
+	// [25:25] SSE1
+	// [26:26] SSE2
+	m_feature_flags = 0x0780a1ff;       // TODO: enable relevant flags here
```
Comment (angelosa) [inline src/devices/cpu/i386/i386.cpp:3547]: Think I've already told you once: enabling features that our p2/p3/p4 cores don't even reach is recipe for disaster.  Not even a Voodoo 1 is exercised outside Arcade machines, 8514 needs work and we don't have a functional AGP infrastructure at all, enabling unhandled SSE1/2 will just annoy anybody genuinely interested in testing these cores.

**Pull request closed without merging**